#881 Allow icons on map layers

tst-mbiancalana opened the pull request:
This commit addresses multiple related things
1) Issue #881, which had errors returning to fusion users when saving a
map layer with an icon chosen for shields or points.  This was caused by
having no image decoders registered in the skia third-party library.

2) The skia library was not being built correctly by SCons. The source
for the skia library exists in a tarball, which is extracted as a step
in the build process, but SCons was trying to gather list of cpp files
up front before the files were extracted. Now the list of files is
gathered from the contents of the tarball.

3) To work around the issue above, the third-party libraries were being
built twice. This worked because on the second round, the cpp files
existed on disk at the right time. This second round is no longer
necessary.

tst-mbiancalana committed:
Allow icons on map layers

This commit addresses multiple related things
1) Issue #881, which had errors returning to fusion users when saving a
map layer with an icon chosen for shields or points.  This was caused by
having no image decoders registered in the skia third-party library.

2) The skia library was not being built correctly by SCons. The source
for the skia library exists in a tarball, which is extracted as a step
in the build process, but SCons was trying to gather list of cpp files
up front before the files were extracted. Now the list of files is
gathered from the contents of the tarball.

3) To work around the issue above, the third-party libraries were being
built twice. This worked because on the second round, the cpp files
existed on disk at the right time. This second round is no longer
necessary.

tst-rwildes commented on earth_enterprise/src/third_party/skia/SConscript:
```diff
@@ -147,9 +175,6 @@ sgl_opts_ssse3_objects = map(
-                   % skia_version).abspath
-
 skia_source_files = sgl_opts_files + sgl_opts_ssse3_files + sgl_cpps
 skia_target = '%s/.extract_skia' % current_dir
```
wonder if we need this anymore...  might not hurt but might not need it too.

tst-mbiancalana replied:
I think this is still important in order to keep the ordering of the build steps correct. The library build still must wait for the tarball to be extracted.

tst-rwildes replied:
Scons will wait because it now knows that the source files to the build command is created by this untar command.  If you give scons all the correct information it does the right thing.  When you take short cuts and hide information from it then it will sometimes do the wrong thing because it doesn't know the full details.  But leaving the dot file for now isn't hurting anything.

tst-rwildes reviewed (changes requested):
Over all looks really good.  Have some suggestions that I'd like to be addressed.  Also should work with @tst-twilk as there might be some conflict here.

tst-eclamar commented:
I'm attempting to compile the code and, after 30 second of successful compiling, I get the following error:
```
...
creating pr/Makefile

sent 569,803 bytes  received 675 bytes  1,140,956.00 bytes/sec
total size is 567,247  speedup is 0.99
+ touch /home/eric/earthenterprise/earth_enterprise/src/NATIVE-REL-x86_64/google/protobuf-py/install/lib/python2.7/site-packages/google/__init__.py
+ touch /home/eric/earthenterprise/earth_enterprise/src/NATIVE-REL-x86_64/google/protobuf-py/.install
creating pr/include/Makefile
creating pr/include/md/Makefile
creating pr/include/obsolete/Makefile
creating pr/include/private/Makefile
creating pr/src/Makefile
creating pr/src/io/Makefile
creating pr/src/linking/Makefile
creating pr/src/malloc/Makefile
creating pr/src/md/Makefile
creating pr/src/md/unix/Makefile
creating pr/src/memory/Makefile
creating pr/src/misc/Makefile
creating pr/src/threads/Makefile
creating pr/tests/Makefile
creating pr/tests/dll/Makefile
creating pr/src/pthreads/Makefile
+ touch /home/eric/earthenterprise/earth_enterprise/src/NATIVE-REL-x86_64/third_party/libjs/.configure_libjs
scons: building terminated because of errors.
scons: *** [build] Error 2
scons: building terminated because of errors.
[eric@centos7b earth_enterprise]$ 

```
I've cleaned and rebuilt twice, each with the same error.  (Note: what I switch back to 'master', everything compiles just fine).

tst-mbiancalana committed:
Revert previous changes to top-level SConstruct

The two-stage build is still required for a majority of the third-party
libraries.

tst-mbiancalana committed:
Fix a typo

tst-mbiancalana commented on earth_enterprise/src/third_party/skia/SConscript:
```diff
@@ -18,29 +18,57 @@
+cpp_downselect_re = re.compile(r'^src/utils/[^/]*(?:SkDeferredCanvas|SkLua|SkThreadUtils_pthread_mach|SkThreadUtils_win|SkDumpCanvas|SkCamera)[^/]*\.cpp$')
+
+# Get a listing of the contents of the skia tarball
+tar_p = subprocess.Popen(['tar', 'tzf', skia_source], stdout=subprocess.PIPE)
```
In looking through some more of the third-party libraries, it looks like there aren't too many places where surveying the contents of the tarball beforehand will help us. We'll keep the technique in our pocket, though.

tst-rwildes replied:
Because we are using fake output files (the dot files) to represent the output of the tar operation all over third party scons can not properly really know if the tar file needs to be re-extracted.  If some of all of the source is deleted for whatever reason but scons sees that dot file it will think executing the un-tar command is not needed because all if the output exists.  The more you tell scons the better and more accurate job it can do.

tst-mbiancalana committed:
Update 5.2.4 release notes, resolution for #881

tst-mbiancalana committed:
Remove unused modified source file

The change contained in this version of SkForceLinking.cpp, which is
commenting out one line, is made to the extracted version by scons
during the build. See earth_enterprise/src/third_party/skia/SConscript
for details.

tst-emillergalow reviewed (approved):
Code looks good and I was able to reproduce the bug in master, and verify that the issue is resolved in this branch.  I successfully saved a 2D Vector layer with the icon changed and didn't see the error message.
